yew/src/dev/mod.rs

```diff
@@ -0,0 +1,228 @@
+                        self.ws.send_with_str(&self.message_queue.pop().unwrap());
+                    }
+                }
+                2 | 3 =>
+                {
+                    #[cfg(feature = "web_sys")]
+                    web_sys::console::error_1(
+                        &"Could not open a connection to Yew's developer tools; are they running?"
+                            .into(),
+                    )
+                }
+                _ => panic!("The WebSocket is in an incorrect state."),
+            }
+        }
+    }
+}
+
+impl DebuggerConnection {
+    /// Creates a new connection to the debugger.
+    /// The URL to which the debugger attempts to connect can be configured by setting some environment variables at compile time.
+    /// If you do not set any of these environment variables, the default values are used.
+    /// The following variables are accepted: `YEW_DEBUGGER_CONNECTION_TYPE`, `YEW_DEBUGGER_HOST` and `YEW_DEBUGGER_PORT`.
+    /// * `YEW_DEBUGGER_CONNECTION_TYPE` – either `ws` or `wss`. `ws` is an insecure WebSocket (but this is fine for local development) and `wss` creates a secure WebSocket which can be used for remote debugging. If you choose `wss` you will need to ensure that your server is configured correctly.
+    /// * `YEW_DEBUGGER_HOST` – a domain or IP address where a debug server can be found.
+    /// * `YEW_DEBUGGER_PORT` – the port on which the debugger is operating.
+    pub fn new() -> Self {
+        let ws_url = format!(
+            "{}://{}:{}",
+            match std::option_env!("YEW_DEBUGGER_CONNECTION_TYPE") {
+                Some(ws_mode) => {
+                    match ws_mode {
+                        "ws" => "ws",
+                        "wss" => "wss",
+                        _ => {
+                            panic!("`{}` is not a valid option for the `YEW_DEBUGGER_SECURE_CONNECTION` environment variable. Use either `ws` for an insecure connection or `wss` for a secure connection.", ws_mode);
+                        }
+                    }
+                }
+                None => {
+                    "ws"
+                }
+            },
+            match std::option_env!("YEW_DEBUGGER_HOST") {
+                Some(url) => {
+                    url
+                }
+                None => {
+                    "localhost"
+                }
+            },
+            match std::option_env!("YEW_DEBUGGER_PORT") {
+                Some(port) => {
+                    port
+                }
+                None => {
+                    "8017"
+                }
+            }
+        );
+        Self {
+            #[cfg(feature = "web_sys")]
+            ws: match web_sys::WebSocket::new(&ws_url) {
+                Ok(s) => s,
+                Err(_) => {
+                    web_sys::console::error_1(&"Error: could not open a connection to the DevTools WebSocket. Are you sure the DevTools backend is running?".into());
```

Comment:
This isn't the right place for this error message. I believe `WebSocket` creation will only fail when the port is blocked: https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/WebSocket